File: tests/node/test_diffusion_smoke.mjs
/**
 * diffusion-rs local image generation smoke tests.
 *
 * Gated on the BLAZEN_TEST_DIFFUSION environment variable.
 * Only runs when the native binding is compiled with the `diffusion` feature.
 *
 * Build first:
 *   cd crates/blazen-node && npm install && npm run build -- --features diffusion
 *
 * Run:
 *   BLAZEN_TEST_DIFFUSION=1 node --test tests/node/test_diffusion_smoke.mjs
 *
 * Tests are sync (no async/await) per project convention: napi-rs's tokio
 * handle lifecycle hangs Node's test runner with async tests.
 */

import test from "ava";

import {
  DiffusionProvider,
  JsDiffusionScheduler,
} from "../../crates/blazen-node/index.js";

const BLAZEN_TEST_DIFFUSION = process.env.BLAZEN_TEST_DIFFUSION;

const MODEL_ID = "stabilityai/stable-diffusion-2-1";

const T = BLAZEN_TEST_DIFFUSION ? test : test.skip;

// Test 1: DiffusionOptions plain-object shape + provider getters reflect them.
// JsDiffusionOptions is a TS interface (plain object), not a constructable class.
// The "constructor" path is DiffusionProvider.create(options); resolved values
// are exposed via provider getters.
T("diffusion-rs local image generation · DiffusionProvider.create accepts options and getters reflect resolved values", (t) => {
  // Skip-on-missing: feature-gate the binding itself.
  if (typeof DiffusionProvider !== "function") {
    t.pass("diffusion feature not built");
    return; // not built with diffusion feature
  }
  if (typeof DiffusionProvider.create !== "function") {
    t.pass("diffusion create factory missing");
    return;
  }

  const provider = DiffusionProvider.create({
    modelId: MODEL_ID,
    width: 768,
    height: 768,
    numInferenceSteps: 25,
    guidanceScale: 8.5,
  });

  t.is(provider.width, 768, "width getter should reflect option");
  t.is(provider.height, 768, "height getter should reflect option");
  t.is(provider.numInferenceSteps, 25, "numInferenceSteps getter should reflect option");
  t.is(provider.guidanceScale, 8.5, "guidanceScale getter should reflect option");
});

// Test 1b: Defaults — when options are omitted, provider exposes documented defaults
// (512x512, 20 steps, 7.5 guidance) per the JsDiffusionOptions docstring.
T("diffusion-rs local image generation · DiffusionProvider.create exposes default getters when options are omitted", (t) => {
  if (typeof DiffusionProvider !== "function") {
    t.pass("diffusion feature not built");
    return;
  }
  if (typeof DiffusionProvider.create !== "function") {
    t.pass("diffusion create factory missing");
    return;
  }

  const provider = DiffusionProvider.create({ modelId: MODEL_ID });

  t.is(provider.width, 512, "default width should be 512");
  t.is(provider.height, 512, "default height should be 512");
  t.is(provider.numInferenceSteps, 20, "default numInferenceSteps should be 20");
  t.is(provider.guidanceScale, 7.5, "default guidanceScale should be 7.5");
});

// Test 2: JsDiffusionScheduler enum exposes the documented variants
// (Euler, EulerA, Dpm, Ddim) with string values.
T("diffusion-rs local image generation · JsDiffusionScheduler exposes the documented variants", (t) => {
  // Skip-on-missing: const enums are inlined by tsc but napi-rs re-exports them
  // at runtime. If the export is missing, the binding wasn't built with diffusion.
  if (JsDiffusionScheduler == null) {
    t.pass("JsDiffusionScheduler not exported");
    return;
  }

  t.is(JsDiffusionScheduler.Euler, "euler", "Euler variant value");
  t.is(JsDiffusionScheduler.EulerA, "eulerA", "EulerA variant value");
  t.is(JsDiffusionScheduler.Dpm, "dpm", "Dpm variant value");
  t.is(JsDiffusionScheduler.Ddim, "ddim", "Ddim variant value");
});

// Test 2b: Provider accepts a scheduler option from the enum.
T("diffusion-rs local image generation · DiffusionProvider.create accepts a scheduler option", (t) => {
  if (typeof DiffusionProvider !== "function") {
    t.pass("diffusion feature not built");
    return;
  }
  if (typeof DiffusionProvider.create !== "function") {
    t.pass("diffusion create factory missing");
    return;
  }
  if (JsDiffusionScheduler == null) {
    t.pass("JsDiffusionScheduler not exported");
    return;
  }

  // Construct with each scheduler variant; should not throw.
  const provider = DiffusionProvider.create({
    modelId: MODEL_ID,
    scheduler: JsDiffusionScheduler.EulerA,
  });
  t.truthy(provider, "provider should be constructed with eulerA scheduler");
  t.is(typeof provider.width, "number", "provider should still expose getters");
});

// Test 3: DiffusionProvider class shape.
// The Rust surface today exposes `create` + four getters. There is no `generate`
// method on the JS side (image generation goes through the higher-level Vision
// pipeline / step handlers). Verify the actual shape rather than fabricating one.
T("diffusion-rs local image generation · DiffusionProvider exposes the expected class shape", (t) => {
  if (typeof DiffusionProvider !== "function") {
    t.pass("diffusion feature not built");
    return;
  }

  // Static factory.
  t.is(typeof DiffusionProvider.create, "function", "create should be a static method");

  // Construct a minimal provider to inspect instance shape.
  if (typeof DiffusionProvider.create !== "function") {
    return;
  }
  const provider = DiffusionProvider.create({ modelId: MODEL_ID });

  // Instance getters as documented in index.d.ts.
  t.is(typeof provider.width, "number", "width getter");
  t.is(typeof provider.height, "number", "height getter");
  t.is(typeof provider.numInferenceSteps, "number", "numInferenceSteps getter");
  t.is(typeof provider.guidanceScale, "number", "guidanceScale getter");

  // Provider should be an instance of DiffusionProvider.
  t.truthy(provider instanceof DiffusionProvider, "provider should be an instance of DiffusionProvider");
});

// Test 4: Skip-on-missing pattern at the describe level.
// If BLAZEN_TEST_DIFFUSION is unset, the whole describe block is skipped. Inside
// each test we additionally guard `typeof DiffusionProvider !== "function"` and
// `JsDiffusionScheduler == null` to handle the case where the env var is set but
// the binding wasn't built with the `diffusion` feature.
T("diffusion-rs local image generation · skips gracefully when the diffusion feature is not built", (t) => {
  const featureAvailable =
    typeof DiffusionProvider === "function" &&
    typeof DiffusionProvider.create === "function" &&
    JsDiffusionScheduler != null;

  // This test asserts the skip-on-missing pattern itself: either the feature
  // is fully present, or the bindings simply don't expose it. There is no
  // half-built state.
  if (!featureAvailable) {
    t.truthy(
      typeof DiffusionProvider !== "function" ||
        typeof DiffusionProvider.create !== "function" ||
        JsDiffusionScheduler == null,
      "missing-feature skip path is consistent",
    );
    return;
  }

  t.is(typeof DiffusionProvider, "function", "DiffusionProvider class should be exported");
  t.is(typeof DiffusionProvider.create, "function", "create factory should be exported");
  t.not(JsDiffusionScheduler, null, "JsDiffusionScheduler enum should be exported");
});

File: tests/python/test_diffusion_smoke.py
"""Smoke tests for the local diffusion-rs image generation backend.

These tests verify that the typed PyO3 binding surface for the diffusion
provider is importable and that the data-carrier types behave correctly
without requiring a real diffusion model on disk.

The diffusion backend is behind a Cargo feature gate. If the wheel was not
built with ``--features diffusion``, the entire module is skipped.

Tests that require a real model are gated on
``BLAZEN_TEST_DIFFUSION=1`` plus ``BLAZEN_DIFFUSION_MODEL_ID=<hf-repo-id>``
to avoid surprise multi-GB downloads or load failures in CI.

Run manually:

    BLAZEN_TEST_DIFFUSION=1 \\
    BLAZEN_DIFFUSION_MODEL_ID=stabilityai/stable-diffusion-2-1 \\
        uv run --no-sync pytest tests/python/test_diffusion_smoke.py -v
"""

import os

import pytest

# The full diffusion typed surface lives behind the `diffusion` Cargo feature.
# If the installed wheel was not built with that feature, skip the entire
# module rather than failing with an ImportError.
try:
    from blazen import (
        DiffusionOptions,
        DiffusionProvider,
        DiffusionScheduler,
        ImageRequest,
    )
except ImportError:
    pytest.skip(
        "blazen was not built with the diffusion feature "
        "(rebuild with: maturin develop -m crates/blazen-py/Cargo.toml --features diffusion)",
        allow_module_level=True,
    )

DIFFUSION_ENABLED = os.environ.get("BLAZEN_TEST_DIFFUSION") == "1"
MODEL_ID = os.environ.get("BLAZEN_DIFFUSION_MODEL_ID")

skip_without_flag = pytest.mark.skipif(
    not DIFFUSION_ENABLED,
    reason="BLAZEN_TEST_DIFFUSION=1 not set",
)

skip_without_model = pytest.mark.skipif(
    not MODEL_ID,
    reason="BLAZEN_DIFFUSION_MODEL_ID not set to a HuggingFace repo id",
)


# ---------------------------------------------------------------------------
# Pure typed-surface tests (no model required, always run)
# ---------------------------------------------------------------------------


def test_diffusion_options_typed_kwargs():
    """DiffusionOptions accepts the documented keyword-only fields."""
    opts = DiffusionOptions(
        model_id="stabilityai/stable-diffusion-2-1",
        device="cpu",
        width=768,
        height=768,
        num_inference_steps=25,
        guidance_scale=7.5,
        scheduler=DiffusionScheduler.EulerA,
        cache_dir="/tmp/blazen-diffusion-cache",
    )
    assert opts.model_id == "stabilityai/stable-diffusion-2-1"
    assert opts.device == "cpu"
    assert opts.width == 768
    assert opts.height == 768
    assert opts.num_inference_steps == 25
    # Float comparison with tolerance for f32 round-trip.
    assert abs(opts.guidance_scale - 7.5) < 1e-5
    assert opts.scheduler == DiffusionScheduler.EulerA
    assert opts.cache_dir == "/tmp/blazen-diffusion-cache"


def test_diffusion_options_defaults_are_none_or_default_scheduler():
    """DiffusionOptions with no kwargs has all-None fields except scheduler."""
    opts = DiffusionOptions()
    assert opts.model_id is None
    assert opts.device is None
    assert opts.width is None
    assert opts.height is None
    assert opts.num_inference_steps is None
    assert opts.guidance_scale is None
    assert opts.cache_dir is None
    # scheduler is a non-Optional enum getter; default per docs is EulerA.
    assert opts.scheduler == DiffusionScheduler.EulerA


def test_diffusion_options_setters_mutate_inplace():
    """DiffusionOptions exposes setters for each field."""
    opts = DiffusionOptions()
    opts.model_id = "runwayml/stable-diffusion-v1-5"
    opts.width = 1024
    opts.height = 1024
    opts.num_inference_steps = 50
    opts.scheduler = DiffusionScheduler.Dpm
    assert opts.model_id == "runwayml/stable-diffusion-v1-5"
    assert opts.width == 1024
    assert opts.height == 1024
    assert opts.num_inference_steps == 50
    assert opts.scheduler == DiffusionScheduler.Dpm


def test_diffusion_scheduler_variants_exist():
    """DiffusionScheduler exposes the four expected variants."""
    assert DiffusionScheduler.Euler != DiffusionScheduler.EulerA
    assert DiffusionScheduler.Dpm != DiffusionScheduler.Ddim
    assert DiffusionScheduler.EulerA != DiffusionScheduler.Dpm
    # Round-trip equality.
    assert DiffusionScheduler.Euler == DiffusionScheduler.Euler
    assert DiffusionScheduler.EulerA == DiffusionScheduler.EulerA
    assert DiffusionScheduler.Dpm == DiffusionScheduler.Dpm
    assert DiffusionScheduler.Ddim == DiffusionScheduler.Ddim


def test_diffusion_provider_class_shape():
    """DiffusionProvider is a class with the expected method/getter surface."""
    # generate_image is the async image-gen entry point per blazen.pyi.
    assert hasattr(DiffusionProvider, "generate_image")
    # Resolved-config getters exposed on the provider.
    assert hasattr(DiffusionProvider, "width")
    assert hasattr(DiffusionProvider, "height")
    assert hasattr(DiffusionProvider, "num_inference_steps")
    assert hasattr(DiffusionProvider, "guidance_scale")
    assert hasattr(DiffusionProvider, "scheduler")


def test_diffusion_provider_default_construction():
    """DiffusionProvider() with no options resolves default config getters."""
    # No options means the provider falls back to its documented defaults
    # (512x512, 20 steps, 7.5 guidance, EulerA scheduler) without touching
    # disk or downloading a model.
    provider = DiffusionProvider()
    assert provider.width == 512
    assert provider.height == 512
    assert provider.num_inference_steps == 20
    assert abs(provider.guidance_scale - 7.5) < 1e-5
    assert provider.scheduler == DiffusionScheduler.EulerA


def test_diffusion_provider_constructs_with_options():
    """DiffusionProvider(options=opts) reflects the provided options."""
    opts = DiffusionOptions(
        width=640,
        height=480,
        num_inference_steps=30,
        guidance_scale=8.0,
        scheduler=DiffusionScheduler.Ddim,
    )
    provider = DiffusionProvider(options=opts)
    assert provider.width == 640
    assert provider.height == 480
    assert provider.num_inference_steps == 30
    assert abs(provider.guidance_scale - 8.0) < 1e-5
    assert provider.scheduler == DiffusionScheduler.Ddim


# ---------------------------------------------------------------------------
# Live tests gated behind BLAZEN_TEST_DIFFUSION=1 + a real HF model id
# ---------------------------------------------------------------------------


@skip_without_flag
@skip_without_model
def test_diffusion_provider_constructs_with_real_model_id():
    """DiffusionProvider(options=opts) accepts a real HF repo id without raising."""
    opts = DiffusionOptions(model_id=MODEL_ID)
    provider = DiffusionProvider(options=opts)
    # Resolved width/height should be the documented default (512) when not
    # overridden, even if the upstream engine integration is still in progress.
    assert provider.width == 512
    assert provider.height == 512


@skip_without_flag
@skip_without_model
@pytest.mark.asyncio
async def test_diffusion_provider_generate_image_runs_or_raises_diffusion_error():
    """generate_image either returns an ImageResult or raises DiffusionError.

    The blazen.pyi docstring documents that the upstream engine integration
    is in progress and ``generate_image`` may raise :class:`DiffusionError`.
    This test accepts either outcome so it remains valid as the integration
    lands.
    """
    opts = DiffusionOptions(model_id=MODEL_ID)
    provider = DiffusionProvider(options=opts)

    request = ImageRequest(
        prompt="a small red cube on a white background",
        width=512,
        height=512,
        num_images=1,
    )

    try:
        result = await provider.generate_image(request)
    except Exception as exc:
        # DiffusionError is the documented failure mode while integration
        # is in progress. Any other exception type is unexpected.
        assert type(exc).__name__ == "DiffusionError", (
            f"unexpected exception {type(exc).__name__}: {exc}"
        )
        return

    # If the call succeeded, the result should be a non-None ImageResult.
    assert result is not None
